klippy/extras/load_cell.py

```diff
@@ -3,21 +3,548 @@
 # Copyright (C) 2024 Gareth Farrington <gareth@waves.ky>
 #
 # This file may be distributed under the terms of the GNU GPLv3 license.
+
 from . import hx71x
 from . import ads1220
+from .bulk_sensor import BatchWebhooksClient
+import logging, collections, itertools
+# We want either Python 3's zip() or Python 2's izip() but NOT 2's zip():
+zip_impl = zip
+try:
+    from itertools import izip as zip_impl # python 2.x izip
+except ImportError: # will be Python 3.x
+    pass
+
+# Helper for event driven webhooks (i.e. non polling based data source)
+class WebhooksHelper(object):
+    def __init__(self, printer):
+        self.printer = printer
+        self.client_cbs = []
+        self.webhooks_start_resp = {}
+
+    # send data to clients
+    def send(self, msg):
+        for client_cb in list(self.client_cbs):
+            res = client_cb(msg)
+            if not res:
+                # This client no longer needs updates - unregister it
+                self.client_cbs.remove(client_cb)
+
+    # Client registration
+    def add_client(self, client_cb):
+        self.client_cbs.append(client_cb)
+
+    # Webhooks registration
+    def _add_api_client(self, web_request):
+        whbatch = BatchWebhooksClient(web_request)
+        self.add_client(whbatch.handle_batch)
+        web_request.send(self.webhooks_start_resp)
+
+    def add_mux_endpoint(self, path, key, value, webhooks_start_resp):
+        self.webhooks_start_resp = webhooks_start_resp
+        wh = self.printer.lookup_object('webhooks')
+        wh.register_mux_endpoint(path, key, value, self._add_api_client)
+
+# Adapter for WebhooksHelper that transforms the response using a function
+# Anything that implements the add_client contract can be a message source
+# Outputs to its own clients
+class WebhooksTransformer(WebhooksHelper):
```

Comment:
Can you elaborate on what this does (as well as WebhooksHelper)?

At first glance, it seems a little odd to make a slightly different interface for just loadcell.  If the BatchWebhooksClient code isn't flexible enough, maybe we should expand the webhooks interface and refactor the other users (like ldc1612 and accelerometers).

Reply:
This goes back to your first comment, lets resolve what we want for a data stream structure. If this is still required I can do a refactor.